File: src/token_saver/cache/persistent_cache.py
"""Persistent SQLite storage for Token-Saver file cache.

Enables cache persistence across MCP server restarts and independent CLI sessions.
Uses SQLite with WAL mode for fast, safe concurrent reads and writes.
"""

from __future__ import annotations

import sqlite3
import time
from pathlib import Path

MAX_CACHEABLE_BYTES = 5 * 1024 * 1024  # 5 MB


class PersistentCache:
    """SQLite-backed persistent store for file hash and content cache."""

    def __init__(self, db_path: Path | str | None = None) -> None:
        if db_path is None:
            cache_dir = Path.home() / ".token-saver"
            cache_dir.mkdir(parents=True, exist_ok=True)
            self._db_path = cache_dir / "cache.db"
        else:
            self._db_path = Path(db_path)
            self._db_path.parent.mkdir(parents=True, exist_ok=True)

        self._init_db()

    def _get_connection(self) -> sqlite3.Connection:
        conn = sqlite3.connect(str(self._db_path), timeout=5.0)
        conn.execute("PRAGMA journal_mode=WAL;")
        conn.execute("PRAGMA synchronous=NORMAL;")
        return conn

    def _init_db(self) -> None:
        """Create cache table and indexes if they do not exist."""
        try:
            with self._get_connection() as conn:
                conn.execute(
                    """
                    CREATE TABLE IF NOT EXISTS file_cache (
                        path TEXT PRIMARY KEY,
                        hash TEXT NOT NULL,
                        content TEXT NOT NULL,
                        read_count INTEGER DEFAULT 1,
                        updated_at REAL NOT NULL
                    )
                    """
                )
                conn.execute("CREATE INDEX IF NOT EXISTS idx_cache_updated ON file_cache(updated_at);")
                conn.execute(
                    """
                    CREATE TABLE IF NOT EXISTS symbol_index (
                        project_root TEXT NOT NULL,
                        file_path TEXT NOT NULL,
                        file_hash TEXT NOT NULL,
                        name TEXT NOT NULL,
                        kind TEXT NOT NULL,
                        line INTEGER NOT NULL,
                        signature TEXT NOT NULL,
                        PRIMARY KEY (project_root, file_path, name, line)
                    )
                    """
                )
                conn.execute("CREATE INDEX IF NOT EXISTS idx_symbol_project_name ON symbol_index(project_root, name);")
                conn.execute(
                    "CREATE INDEX IF NOT EXISTS idx_symbol_project_file ON symbol_index(project_root, file_path);"
                )
                conn.execute(
                    """
                    CREATE TABLE IF NOT EXISTS file_index_meta (
                        project_root TEXT NOT NULL,
                        file_path TEXT NOT NULL,
                        file_hash TEXT NOT NULL,
                        mtime REAL NOT NULL,
                        PRIMARY KEY (project_root, file_path)
                    )
                    """
                )
        except Exception:
            pass

    def get_entry(self, path: str) -> tuple[str, str, int] | None:
        """Fetch cached entry (hash, content, read_count) for a given path."""
        try:
            with self._get_connection() as conn:
                cursor = conn.cursor()
                cursor.execute(
                    "SELECT hash, content, read_count FROM file_cache WHERE path = ?",
                    (path,),
                )
                row = cursor.fetchone()
                if row:
                    return str(row[0]), str(row[1]), int(row[2])
                return None
        except Exception:
            return None

    def set_entry(self, path: str, hash_val: str, content: str, read_count: int = 1) -> None:
        """Store or update a cached file entry. Protects SQLite from giant files > 5MB."""
        try:
            stored_content = content
            if len(content) > MAX_CACHEABLE_BYTES:
                stored_content = f"__TOKEN_SAVER_LARGE_FILE__:{len(content)}"

            with self._get_connection() as conn:
                conn.execute(
                    """
                    INSERT INTO file_cache (path, hash, content, read_count, updated_at)
                    VALUES (?, ?, ?, ?, ?)
                    ON CONFLICT(path) DO UPDATE SET
                        hash = excluded.hash,
                        content = excluded.content,
                        read_count = excluded.read_count,
                        updated_at = excluded.updated_at
                    """,
                    (path, hash_val, stored_content, read_count, time.time()),
                )
        except Exception:
            pass

    def invalidate(self, path: str) -> None:
        """Remove an entry from the persistent cache."""
        try:
            with self._get_connection() as conn:
                conn.execute("DELETE FROM file_cache WHERE path = ?", (path,))
        except Exception:
            pass

    def clear(self) -> None:
        """Clear all entries in the persistent cache."""
        try:
            with self._get_connection() as conn:
                conn.execute("DELETE FROM file_cache")
        except Exception:
            pass

    def prune(self, max_entries: int = 5000, max_age_days: int = 7) -> int:
        """Prune old entries to prevent unbounded disk usage."""
        try:
            cutoff = time.time() - (max_age_days * 86400)
            with self._get_connection() as conn:
                cursor = conn.cursor()
                cursor.execute("DELETE FROM file_cache WHERE updated_at < ?", (cutoff,))
                deleted = cursor.rowcount

                cursor.execute("SELECT COUNT(*) FROM file_cache")
                count = cursor.fetchone()[0]
                if count > max_entries:
                    excess = count - max_entries
                    cursor.execute(
                        """
                        DELETE FROM file_cache WHERE path IN (
                            SELECT path FROM file_cache ORDER BY updated_at ASC LIMIT ?
                        )
                        """,
                        (excess,),
                    )
                    deleted += cursor.rowcount
                return deleted
        except Exception:
            return 0

    def count_entries(self) -> int:
        """Return total number of cached entries."""
        try:
            with self._get_connection() as conn:
                cursor = conn.cursor()
                cursor.execute("SELECT COUNT(*) FROM file_cache")
                row = cursor.fetchone()
                return int(row[0]) if row else 0
        except Exception:
            return 0

    def get_indexed_file_meta(self, project_root: str, file_path: str) -> tuple[str, float] | None:
        """Fetch indexed file hash and mtime for a file within a project."""
        try:
            with self._get_connection() as conn:
                cursor = conn.cursor()
                cursor.execute(
                    "SELECT file_hash, mtime FROM file_index_meta WHERE project_root = ? AND file_path = ?",
                    (project_root, file_path),
                )
                row = cursor.fetchone()
                if row:
                    return str(row[0]), float(row[1])
                return None
        except Exception:
            return None

    def set_file_symbols(
        self,
        project_root: str,
        file_path: str,
        file_hash: str,
        mtime: float,
        symbols: list[dict],
    ) -> None:
        """Atomically update the symbol index and metadata for a specific file."""
        try:
            with self._get_connection() as conn:
                conn.execute(
                    "DELETE FROM symbol_index WHERE project_root = ? AND file_path = ?",
                    (project_root, file_path),
                )
                for s in symbols:
                    conn.execute(
                        """
                        INSERT OR REPLACE INTO symbol_index (
                            project_root, file_path, file_hash, name, kind, line, signature
                        ) VALUES (?, ?, ?, ?, ?, ?, ?)
                        """,
                        (
                            project_root,
                            file_path,
                            file_hash,
                            s.get("name", ""),
                            s.get("kind", ""),
                            s.get("line", 0),
                            s.get("signature", ""),
                        ),
                    )
                conn.execute(
                    """
                    INSERT INTO file_index_meta (project_root, file_path, file_hash, mtime)
                    VALUES (?, ?, ?, ?)
                    ON CONFLICT(project_root, file_path) DO UPDATE SET
                        file_hash = excluded.file_hash,
                        mtime = excluded.mtime
                    """,
                    (project_root, file_path, file_hash, mtime),
                )
        except Exception:
            pass

    def get_file_symbols(self, project_root: str, file_path: str) -> list[dict]:
        """Fetch all indexed symbols for a specific file."""
        try:
            with self._get_connection() as conn:
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT name, kind, file_path, line, signature, file_hash
                    FROM symbol_index
                    WHERE project_root = ? AND file_path = ?
                    ORDER BY line ASC
                    """,
                    (project_root, file_path),
                )
                rows = cursor.fetchall()
                return [
                    {
                        "name": r[0],
                        "kind": r[1],
                        "file_path": r[2],
                        "line": int(r[3]),
                        "signature": r[4],
                        "file_hash": r[5],
                    }
                    for r in rows
                ]
        except Exception:
            return []

    def search_symbols(
        self,
        project_root: str,
        query: str,
        exact: bool = False,
        max_results: int = 15,
    ) -> list[dict]:
        """Search symbols scoped strictly to a project_root using SQLite indexes."""
        try:
            with self._get_connection() as conn:
                cursor = conn.cursor()
                if exact:
                    cursor.execute(
                        """
                        SELECT name, kind, file_path, line, signature, file_hash
                        FROM symbol_index
                        WHERE project_root = ? AND name = ?
                        LIMIT ?
                        """,
                        (project_root, query, max_results),
                    )
                else:
                    like_pattern = f"%{query}%"
                    cursor.execute(
                        """
                        SELECT name, kind, file_path, line, signature, file_hash
                        FROM symbol_index
                        WHERE project_root = ? AND name LIKE ?
                        LIMIT ?
                        """,
                        (project_root, like_pattern, max_results),
                    )
                rows = cursor.fetchall()
                return [
                    {
                        "name": r[0],
                        "kind": r[1],
                        "file_path": r[2],
                        "line": int(r[3]),
                        "signature": r[4],
                        "file_hash": r[5],
                    }
                    for r in rows
                ]
        except Exception:
            return []

    def get_all_project_symbols(self, project_root: str) -> list[dict]:
        """Return all indexed symbols for a specific project."""
        try:
            with self._get_connection() as conn:
                cursor = conn.cursor()
                cursor.execute(
                    """
                    SELECT name, kind, file_path, line, signature, file_hash
                    FROM symbol_index
                    WHERE project_root = ?
                    ORDER BY file_path ASC, line ASC
                    """,
                    (project_root,),
                )
                rows = cursor.fetchall()
                return [
                    {
                        "name": r[0],
                        "kind": r[1],
                        "file_path": r[2],
                        "line": int(r[3]),
                        "signature": r[4],
                        "file_hash": r[5],
                    }
                    for r in rows
                ]
        except Exception:
            return []

    def prune_project_deleted_files(self, project_root: str, valid_file_paths: set[str]) -> int:
        """Remove symbols and metadata for files that no longer exist in the project."""
        try:
            with self._get_connection() as conn:
                cursor = conn.cursor()
                cursor.execute(
                    "SELECT file_path FROM file_index_meta WHERE project_root = ?",
                    (project_root,),
                )
                stored_files = [r[0] for r in cursor.fetchall()]
                deleted_count = 0
                for f in stored_files:
                    if f not in valid_file_paths:
                        conn.execute(
                            "DELETE FROM symbol_index WHERE project_root = ? AND file_path = ?",
                            (project_root, f),
                        )
                        conn.execute(
                            "DELETE FROM file_index_meta WHERE project_root = ? AND file_path = ?",
                            (project_root, f),
                        )
                        deleted_count += 1
                return deleted_count
        except Exception:
            return 0

File: crates/token-saver-core/src/cache/persistent_cache.rs
//! L2 Persistent SQLite Cache for Token-Saver.
//!
//! Stores AST digests and symbol indexes across sessions and IDE reboots in ~/.token-saver/cache.db.

use std::path::PathBuf;
use rusqlite::{params, Connection, Result};
use crate::models::IndexedSymbol;

pub struct PersistentCache {
    db_path: PathBuf,
}

impl PersistentCache {
    pub fn new() -> Result<Self> {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
        let dir = home.join(".token-saver");
        std::fs::create_dir_all(&dir).ok();
        let db_path = dir.join("cache.db");

        let cache = Self { db_path };
        cache.init_schema()?;
        Ok(cache)
    }

    pub fn with_path(db_path: PathBuf) -> Result<Self> {
        let cache = Self { db_path };
        cache.init_schema()?;
        Ok(cache)
    }

    fn connect(&self) -> Result<Connection> {
        let conn = Connection::open(&self.db_path)?;
        conn.execute_batch(
            "PRAGMA journal_mode = WAL;
             PRAGMA synchronous = NORMAL;
             PRAGMA foreign_keys = ON;",
        )?;
        Ok(conn)
    }

    fn init_schema(&self) -> Result<()> {
        let conn = self.connect()?;
        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS file_cache (
                file_path TEXT PRIMARY KEY,
                content_hash TEXT NOT NULL,
                mtime REAL NOT NULL,
                token_count INTEGER NOT NULL,
                cached_at INTEGER NOT NULL
            );

            CREATE TABLE IF NOT EXISTS symbol_index (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                project_root TEXT NOT NULL,
                file_path TEXT NOT NULL,
                name TEXT NOT NULL,
                kind TEXT NOT NULL,
                line INTEGER NOT NULL,
                signature TEXT NOT NULL,
                file_hash TEXT NOT NULL,
                mtime REAL NOT NULL
            );

            CREATE INDEX IF NOT EXISTS idx_sym_proj_name ON symbol_index(project_root, name);
            CREATE INDEX IF NOT EXISTS idx_sym_proj_file ON symbol_index(project_root, file_path);",
        )?;

        // Auto-migration: ensure mtime column exists if table was created in earlier beta
        let _ = conn.execute("ALTER TABLE symbol_index ADD COLUMN mtime REAL DEFAULT 0.0", []);

        Ok(())
    }

    pub fn search_symbols(
        &self,
        project_root: &str,
        query: &str,
        exact: bool,
        max_results: usize,
    ) -> Result<Vec<IndexedSymbol>> {
        let conn = self.connect()?;
        let mut symbols = Vec::new();

        if exact {
            let mut stmt = conn.prepare(
                "SELECT name, kind, file_path, line, signature, file_hash
                 FROM symbol_index
                 WHERE project_root = ? AND name = ?
                 LIMIT ?",
            )?;
            let rows = stmt.query_map(params![project_root, query, max_results], |row| {
                Ok(IndexedSymbol {
                    name: row.get(0)?,
                    kind: row.get(1)?,
                    file_path: row.get(2)?,
                    line: row.get(3)?,
                    signature: row.get(4)?,
                    content_hash: row.get(5)?,
                })
            })?;
            for s in rows {
                symbols.push(s?);
            }
        } else {
            let pattern = format!("%{query}%");
            let mut stmt = conn.prepare(
                "SELECT name, kind, file_path, line, signature, file_hash
                 FROM symbol_index
                 WHERE project_root = ? AND name LIKE ?
                 LIMIT ?",
            )?;
            let rows = stmt.query_map(params![project_root, pattern, max_results], |row| {
                Ok(IndexedSymbol {
                    name: row.get(0)?,
                    kind: row.get(1)?,
                    file_path: row.get(2)?,
                    line: row.get(3)?,
                    signature: row.get(4)?,
                    content_hash: row.get(5)?,
                })
            })?;
            for s in rows {
                symbols.push(s?);
            }
        }

        Ok(symbols)
    }

    pub fn set_file_symbols(
        &self,
        project_root: &str,
        file_path: &str,
        file_hash: &str,
        mtime: f64,
        symbols: &[IndexedSymbol],
    ) -> Result<()> {
        let mut conn = self.connect()?;
        let tx = conn.transaction()?;

        // Delete existing symbols for this file
        tx.execute(
            "DELETE FROM symbol_index WHERE project_root = ? AND file_path = ?",
            params![project_root, file_path],
        )?;

        // Insert fresh symbols
        {
            let mut stmt = tx.prepare(
                "INSERT INTO symbol_index (project_root, file_path, name, kind, line, signature, file_hash, mtime)
                 VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            )?;
            for s in symbols {
                stmt.execute(params![
                    project_root,
                    file_path,
                    s.name,
                    s.kind,
                    s.line,
                    s.signature,
                    file_hash,
                    mtime
                ])?;
            }
        }

        tx.commit()?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_persistent_cache_symbols() {
        let temp = tempfile::NamedTempFile::new().unwrap();
        let cache = PersistentCache::with_path(temp.path().to_path_buf()).unwrap();

        let syms = vec![IndexedSymbol {
            name: "calculate_tokens".to_string(),
            kind: "function".to_string(),
            file_path: "src/counter.rs".to_string(),
            line: 42,
            signature: "pub fn calculate_tokens(s: &str) -> usize".to_string(),
            content_hash: "hash123".to_string(),
        }];

        cache.set_file_symbols("/test_root", "src/counter.rs", "hash123", 1000.0, &syms).unwrap();

        // Exact search
        let exact = cache.search_symbols("/test_root", "calculate_tokens", true, 10).unwrap();
        assert_eq!(exact.len(), 1);
        assert_eq!(exact[0].name, "calculate_tokens");

        // Fuzzy/substring search
        let fuzzy = cache.search_symbols("/test_root", "tokens", false, 10).unwrap();
        assert_eq!(fuzzy.len(), 1);
    }
}
